### PR TITLE
[Moments] Add MomentsClientFactory

### PR DESCRIPTION
**Pull request by nacho9900**
## Description

Most of the time we want to use this configuration of the `MomentsClient` and we duplicate this code a lot.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [X] I have read the [CONTRIBUTING](https://github.com/poap-xyz/poap.js/blob/main/documentation/CONTRIBUTING.md) document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+const getAuthenticationProvider = (
+  params: GetMomentsDefaultClientOptions,
+): AuthenticationProvider => {
+  if (params.moments.authenticationProvider) {
+    return params.moments.authenticationProvider;
+  }
+
+  if (!params.moments.oAuthCredentials) {
+    throw new Error(
```
Comment (rlajous) [inline packages/moments/src/client/getMomentsDefaultClient.ts:43]: Can we make this a typed error?

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+interface GetMomentsDefaultClientOptions {
+  moments: {
+    momentsBaseUrl?: string;
+    /**
+     * The authentication provider used for authenticating requests to the moments API.
+     * If this is provided, oAuthAuthentication will be ignored.
+     */
+    authenticationProvider?: AuthenticationProvider;
+    /**
+     * The OAuth authentication configuration used for authenticating requests to the moments API.
```
Comment (rlajous) [inline packages/moments/src/client/getMomentsDefaultClient.ts:18]: Can you specify this in a js doc of the interface instead of here?

**Review by rlajous — COMMENTED** [review]
You are missing a version change

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  });
+
+  const authenticationProvider = getAuthenticationProvider(params);
+
+  const momentsAPI = new PoapMomentsApi({
+    baseUrl: params.moments.momentsBaseUrl,
+    authenticationProvider,
+  });
+
+  return new MomentsClient(momentsAPI, compassProvider);
```
Comment (reobin) [inline packages/moments/src/client/getMomentsDefaultClient.ts:70]: This whole helper function seems very helpful in pretty much all scenarios. I wonder if a static method on the MomentsClient class would be even better to help this method get known by other devs.

File: packages/moments/src/client/getMomentsDefaultClient.ts (hunk 1 above)
Comment (jm42) [reply packages/moments/src/client/getMomentsDefaultClient.ts:43]: Is this errors supposed to be catched and do some business depending on it? If not I don't see why it needs to be types. Is an error for the developer to see and fix his/her configuration.

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    params.moments.oAuthCredentials.clientSecret,
+    params.moments.oAuthCredentials.oAuthServerDomain ?? DEFAULT_OAUTH_SERVER,
+  );
+};
+
+const getMomentsDefaultClient = (
+  params: GetMomentsDefaultClientOptions,
+): MomentsClient => {
+  const compassProvider = new PoapCompass({
+    apiKey: params.compass?.compassApiKey ?? '',
```
Comment (jm42) [inline packages/moments/src/client/getMomentsDefaultClient.ts:59]: I don't think the api key should be given an empty string when the provider requires it.

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+     * If this is provided, it will be used to create an `AuthenticationProviderOAuth` instance.
+     */
+    oAuthCredentials?: {
+      clientId: string;
+      clientSecret: string;
+      oAuthServerDomain?: string;
+    };
+  };
+  compass?: {
+    compassApiKey?: string;
```
Comment (jm42) [inline packages/moments/src/client/getMomentsDefaultClient.ts:28]: The API key in compass provider is required.

```suggestion
  compass: {
    compassApiKey: string;
```

File: packages/moments/src/client/getMomentsDefaultClient.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * The OAuth authentication configuration used for authenticating requests to the moments API.
+     * If this is provided, it will be used to create an `AuthenticationProviderOAuth` instance.
+     */
+    oAuthCredentials?: {
+      clientId: string;
+      clientSecret: string;
+      oAuthServerDomain?: string;
+    };
+  };
```
Comment (jm42) [inline packages/moments/src/client/getMomentsDefaultClient.ts:26]: You can do union to enforce one or the other with types.

Something like:

```
  moments: { momentsBaseUrl?: string; } & (
    | { authenticationProvider: ... }
    | { oAuthCredentials: { ... }
  );
```

File: packages/moments/src/client/getMomentsDefaultClient.ts (hunk 1 above)
Comment (nacho9900) [reply packages/moments/src/client/getMomentsDefaultClient.ts:43]: Agree with Juan

File: packages/moments/src/client/getMomentsDefaultClient.ts (hunk 6 above)
Comment (nacho9900) [reply packages/moments/src/client/getMomentsDefaultClient.ts:26]: Thanks!

**Review by jm42 — APPROVED** [review]

File: packages/moments/src/client/getMomentsDefaultClient.ts (hunk 3 above)
Comment (reobin) [reply packages/moments/src/client/getMomentsDefaultClient.ts]: Oh cool. If it's a common pattern people will look for, then that's fine.

**Review by reobin — APPROVED** [review]

**Review by jm42 — APPROVED** [review]

**Review by rlajous — COMMENTED** [review]
Could you let me know if you created a test version to see if it worked? 

- [ ] Mark this if done.

**Pull request merged**